### PR TITLE
nixos/systemd: allow running shellcheck on generated systemd scripts

### PR DESCRIPTION
**Pull request by r-vdp**
## Description of changes

With this PR we can optionally run shellcheck on generated systemd scripts.
Our nixos closure already depends on GHC (and shellcheck), so this does not affect the runtime closure.

~~I fixed the scripts that I use on my machines, but there is definitely a lot more stuff to fix before this could become the default.~~
I will submit these as separate PRs.

For now, this sits behind an enable-option. It would be great if some people would enable this and submit fixes for existing scripts.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/lib/systemd-lib.nix
Comment (SuperSandro2000) [inline nixos/lib/systemd-lib.nix:344]: Naive question: Should set set -eou pipefail?

File: nixos/lib/systemd-lib.nix
Comment (SuperSandro2000) [inline nixos/lib/systemd-lib.nix:358]: nice 👍🏼

File: nixos/modules/virtualisation/nixos-containers.nix
Comment (SuperSandro2000) [inline nixos/modules/virtualisation/nixos-containers.nix:94]: I am not sure what the layout is before this. If $root already exists, we could also drop the -p where possible.

Also I don't think your change is fixing the issue shellcheck is noting. The directories along the path still will have the same permissions as before and the ones created by -p still have the default umask permissions.

File: nixos/modules/virtualisation/nixos-containers.nix
Comment (SuperSandro2000) [inline nixos/modules/virtualisation/nixos-containers.nix:163]: Since extra flags is a string here, this will work. Normally, especially in stdenv, extraFlags is an array, which would mean we would mess with quoting here.

File: nixos/modules/system/activation/activation-script.nix
Comment (SuperSandro2000) [inline nixos/modules/system/activation/activation-script.nix:171]: We would add /empty to the beginning of the path here which I don't think we want to do.

```suggestion
          export PATH=
```

File: nixos/modules/virtualisation/ec2-data.nix
Comment (SuperSandro2000) [inline nixos/modules/virtualisation/ec2-data.nix:56]: I think we are always using bash here, right? Then we might as well use [[

File: nixos/modules/security/acme/default.nix
Comment (SuperSandro2000) [inline nixos/modules/security/acme/default.nix:170]: good catch 👍🏼

File: nixos/modules/tasks/filesystems/zfs.nix
Comment (SuperSandro2000) [inline nixos/modules/tasks/filesystems/zfs.nix:65]: We should add a short comment, why this was added

**Review by SuperSandro2000 — COMMENTED** [review]
Very good idea. I think the only thing that really needs some discussion, is what we are going to do about the shellcheck mkdir -p error. Maybe we ignore it by default for now?

Also please remind me in a week or so, that I want to test this on my deployments.

**Review by ElvishJerricco — DISMISSED** [review]
Seems like this should be something that can be enabled per-unit, with the global setting acting as the default.

File: nixos/lib/systemd-lib.nix
Comment (r-vdp) [reply nixos/lib/systemd-lib.nix:344]: Yeah but that can be done in another PR then. It can also cause breakage, especially `-u`.

**Comment by r-vdp** [comment]
> Seems like this should be something that can be enabled per-unit, with the global setting acting as the default.

@ElvishJerricco that's already how it's done. There is a global switch (defaults to `false` for now) and a per-unit switch that defaults to the value of the global switch.

So it's possible to enable or disable this on a per-unit basis, independent of the global option.

File: nixos/lib/systemd-lib.nix
Comment (SuperSandro2000) [reply nixos/lib/systemd-lib.nix:344]: Yeah, that's true

**Comment by ElvishJerricco** [comment]
Ah, I misunderstood what I was reading. Great!

File: nixos/modules/virtualisation/nixos-containers.nix
Comment (r-vdp) [reply nixos/modules/virtualisation/nixos-containers.nix:163]: So there's nothing to be done here? Would you do this differently?

For an array, I guess you'd want to do `"${extraFlags[@]}"` instead, but then we'd need to test first whether the variable is an array or a string and act accordingly. I'm not sure that that's needed here currently?

**Review by ElvishJerricco — CHANGES_REQUESTED** [review]

File: nixos/modules/tasks/filesystems/zfs.nix
Comment (r-vdp) [reply nixos/modules/tasks/filesystems/zfs.nix:880]: @phaer I think we don't need the quotes around the nix interpolation, do we? It should be sufficient to properly quote the generated bash code.

Could you test if that works? I think you made this change because of your zfs config, right?

File: nixos/modules/tasks/filesystems/zfs.nix
Comment (r-vdp) [reply nixos/modules/tasks/filesystems/zfs.nix]: I rewrote this part such that we no longer need to ignore this warning.

File: nixos/modules/system/activation/activation-script.nix
Comment (r-vdp) [reply nixos/modules/system/activation/activation-script.nix]: Done. The same is already done on line 38 in the same file, not sure why exactly.

File: nixos/modules/virtualisation/ec2-data.nix
Comment (SuperSandro2000) [reply nixos/modules/virtualisation/ec2-data.nix:56]: > it may be slightly nicer to stick to posix?

Probably not, using [ wrong is pretty easy and [[ avoids some quoting traps.

File: nixos/modules/virtualisation/nixos-containers.nix
Comment (SuperSandro2000) [reply nixos/modules/virtualisation/nixos-containers.nix:163]: yeah, unfortunate naming

File: nixos/modules/tasks/filesystems/zfs.nix
Comment (phaer) [reply nixos/modules/tasks/filesystems/zfs.nix:880]: Sorry, I only saw this notification now. I think your assumption is correct; thanks for the catch @ElvishJerricco. Its just that I am in a train right now and won't be able to properly test this on a zfs machine for a good week, when I return home

Comment (r-vdp) [reply nixos/modules/tasks/filesystems/zfs.nix]: I dropped this commit in order to unblock this PR, since I don't know zfs well enough to test this.

**Comment by r-vdp** [comment]
Any other remarks on this? I will still add a changelog entry in any case.

**Review by emilazy — APPROVED** [review]
ACME changes LGTM. Didn’t review anything else.

**Review by ElvishJerricco — DISMISSED** [review]
***Whoa***. Upon review, there are *a ton of changes* that have nothing to do with the logic of this PR. I will not review that. Please remove those changes and open a separate PR for them.

**Comment by emilazy** [comment]
One PR per service or group of services is probably reasonable since they can be reviewed separately by the people responsible for them.

**Comment by r-vdp** [comment]
OK, sure, they are already split in different commits, I'll make separate PRs for them.

**Comment by r-vdp** [comment]
I removed the fixes for different services. I will make separate PRs for these.

You won't be able to build a NixOS system with shellcheck for services enabled until we also merge fixes for some of those.

**Review by Mindavi — APPROVED** [review]
I think splitting this up was the right call and on surface this looks good to me and I ACK the idea too.

File: nixos/modules/virtualisation/nixos-containers.nix
Comment (r-vdp) [reply nixos/modules/virtualisation/nixos-containers.nix]: This was pulled out of this PR.

**Review by phaer — APPROVED** [review]
Very nice feature! Code looks good to me and feature works as intended. PR quite easy to review IMO.

**Pull request merged**
